fookb/fookb.1.in
.\" Copyright (c) 1998-2004 Alexey Vyskubov <alexey@mawhrin.net>
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the Artistic License
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
.TH fookb 1 "September 2017"
.SH NAME
fookb \- Xkb state indicator with optional Window Maker support
.SH SYNOPSIS
.B fookb [options]

.SH DESCRIPTION
\fBfookb\fP is a Xkb state indicator. It displays
the icon corresponding to the number of Xkb locked group. You may use
mouse-clicks to change Xkb locked group with mouse. Additionally,
\fBfookb\fP can be configured to execute a program on changing group.
.P
\fBfookb\fP man page is based on \fBWMMail.app\fP man page. Program
was written from scratch.
.P
To use \fBfookb\fP with the Window Maker Dock, simply drag the
\fBfookb\fP icon to the Window Maker Dock (if \fBfookb\fP was compiled
with Window Maker support).

.SH OPTIONS
.TP
NB! Command line parameters take precedence over config file!
.TP
All the icons should be of the same size or fookb will refuse to run.
fookb will warn you if you use icons of size more than 64x64, but it will try
to run anyway.
.TP
.B --icon1
Icon to show for the 1st Xkb group
.TP
.B --icon2
Icon to show for the 2nd Xkb group
.TP
.B --icon3
Icon to show for the 3rd Xkb group
.TP
.B --icon4
Icon to show for the 4th Xkb group
.TP
.B --iconboom
Icon to show when Xkb system goes crazy
.TP
.B -d,--display
X display to use (normally not needed)

.SH CONFIGURATION FILE
The configuration file consist of a single PropList dictionary, which
in turn is composed of several PropList key\-value pairs. The
recognized keys is:
.TP
.B Icon1 Filename (mandatory)
XPM file contains an icon for 1st XKB group.
.TP
.B Icon2 Filename (mandatory)
XPM file contains an icon for 2nd XKB group.
.TP
.B Icon3 Filename (mandatory)
XPM file contains an icon for 3rd XKB group.
.TP
.B Icon4 Filename (mandatory)
XPM file contains an icon for 4th XKB group.
.TP
.B IconBoom Filename (mandatory)
XPM file contains an icon for "broken Xkb". Please look in
BUGS section.
.TP
.B Sound (Yes|No)
\fBfookb\fP will run command from "Command" value (see below), if this
parameter is set to "Yes". You'll receive (non-fatal) error message if
this parameter is not defined. In any other case command will not be
run.
.TP
.B Command command_to_run
This command will be run if XKB locked group is changed and "Sound" is
set to "Yes".

.SH EXAMPLES
Here is an example of configuration file.
.P
.nf
{
  Icon1 = "/usr/local/share/fookb/icon1.xpm";
  Icon2 = "/usr/local/share/fookb/icon2.xpm";
  Icon3 = "/usr/local/share/fookb/icon3.xpm";
  Icon4 = "/usr/local/share/fookb/icon4.xpm";
  IconBoom = "/usr/local/share/fookb/crash.xpm";
  Sound = "Yes";
  Command = "/usr/bin/play /usr/local/share/fookb/click.au";
}
.fi

.SH FILES
.TP
.B ~/.fookb
\fBfookb\fP defaults

.SH BUGS
Sometimes \fBfookb\fP receives message about changing XKB group, but
message.state.group does not point to any existing XKB group. In this
case \fBfookb\fP will show IconBoom (from config file) picture.
.P
As far as I know the only way to look at this bug is to run vmware.

.SH VERSION
This man page is [hopefully] up-to-date for version @PACKAGE_VERSION@ of \fBfookb\fP.

.SH AUTHORS
.P
\fBfookb\fP is (c) 1998-2004 Alexey Vyskubov <alexey@mawhrin.net>.
.P
This manpage was written by Alexey Vyskubov <alexey@mawhrin.net>.


fookb/fookb.c
/*
 * fookb.c
 *
 * (c) 1998-2004 Alexey Vyskubov <alexey@mawhrin.net>
 */

#include <stdlib.h>		/* malloc() */
#include <stdio.h>		/* puts() */

/* X Window headers */
#include <X11/Xlib.h>
#include <X11/Xutil.h>

/* XKB fun */
#include <X11/XKBlib.h>

#include <libdockapp/dockapp.h>

/* My own fun */
#include "fookb.h"
#include "images.h"
#include "sound.h"
#include "opts.h"

char *icon1 = NULL;
char *icon2 = NULL;
char *icon3 = NULL;
char *icon4 = NULL;
char *iconboom = NULL;
char *display = NULL;

#define sterror(x) (void)printf("Strange error, please report! %s:%d, %s\n",\
		__FILE__, __LINE__, x)

int main(int argc, register char *argv[])
{
	XkbEvent labuda;        /* Xkb event. X event will be labuda.core */
	int state = 0;		/* We suppose that latin keyboard is the
				   primal state FIXME */
	Pixmap pixmap;
	DAProgramOption options[] = {
		{NULL, "--icon1",
		 "Icon to show for the 1st Xkb group",
		 DOString, False, {&icon1}},
		{NULL, "--icon2",
		 "Icon to show for the 2nd Xkb group",
		 DOString, False, {&icon2}},
		{NULL, "--icon3",
		 "Icon to show for the 3rd Xkb group",
		 DOString, False, {&icon3}},
		{NULL, "--icon4",
		 "Icon to show for the 4th Xkb group",
		 DOString, False, {&icon4}},
		{NULL, "--iconboom",
		 "Icon to show when Xkb system goes crazy",
		 DOString, False, {&iconboom}},
		{"-d", "--display",
		 "X display to use (normally not needed)",
		 DOString, False, {&display}},
	};


	DAParseArguments(argc, argv, options, 6,
			 "XKB state indicator for Window Maker",
			 PACKAGE_STRING);

	DAOpenDisplay(display, argc, argv);
	read_images(DADisplay);		/* Let's read icon images */
	DACreateIcon(PACKAGE_NAME, get_width(), get_height(), argc, argv);
	XSelectInput(DADisplay, DAWindow, ButtonPressMask);

/* We would like receive the only Xkb event: XkbStateNotify. And only
 * when XkbLockGroup happens. */

	if (False == XkbSelectEvents(DADisplay,
			XkbUseCoreKbd,
			XkbAllEventsMask,
			0)) {
		sterror("Cannot XkbSelectEvents. It's your problem -- not mine.");
		exit(EXIT_FAILURE);
	} /* Deselect all events */

	if (False == XkbSelectEventDetails(DADisplay,
				XkbUseCoreKbd,
				XkbStateNotify,
				XkbAllEventsMask,
				XkbGroupLockMask)) {
		sterror("Cannot XkbSelectEventDetails. It's your problem -- not mine.");
		exit(EXIT_FAILURE);
	} /* Select XkbStateNotify/XkbgroupLock */

	pixmap = DAMakePixmap();

	update_window(pixmap, DAGC, state, DADisplay);
	DASetPixmap(pixmap);

	DAShow();

/* HELLO! HELLO! HELLO! Is that our GOOD FRIEND main loop here? */
	while (1) {
		XNextEvent(DADisplay, &labuda.core);
		switch (labuda.core.type) {

		case ButtonPress:

			switch (labuda.core.xbutton.button) {
			case Button1:
#ifdef DEBUG
				puts("Button1 pressed.");
#endif
				XkbLockGroup(DADisplay,
					     XkbUseCoreKbd,
					     (state + 1) % 4);
				break;
			case Button2:
#ifdef DEBUG
				puts("Button2 pressed.");
#endif
				XkbLockGroup(DADisplay,
					     XkbUseCoreKbd,
					     (state + 3) % 4);
				break;
			case Button3:
#ifdef DEBUG
				puts("Button3 pressed, bye.");
#endif
				XFreeGC(DADisplay, DAGC);
				XDestroyWindow(DADisplay, DAWindow);
				XCloseDisplay(DADisplay);
				exit(0);
			}
			break;
		default:	/* XkbLockGroup happens : FIXME */
			drip();
			state = labuda.state.group;
#ifdef DEBUG
			printf("%u\n", state);
#endif
			if ((state < 0) || (state > 4))
				state = 4;
			update_window(pixmap, DAGC, state, DADisplay);
			DASetPixmap(pixmap);

#ifdef DEBUG
			puts(".");	/* XkbLockGroup happens */
#endif
		}
	}
}
